### ISSUE TITLE
Use DP+FSDP device mesh dimensions for scaling loss with default value of average_tokens_across_devices: True

### ISSUE THREAD

**Issue by winglian** [huggingface/transformers#39648]
### System Info

So with [this PR changing the default for how the train loss is scaled](https://github.com/huggingface/transformers/pull/39395) and with [accelerate now supporting composable parallelisms](https://github.com/huggingface/accelerate/pull/3682), we need to update these token scaling to only scale across the data parallel dimension groups/meshes and ignore any TP or CP meshes. 

https://github.com/huggingface/transformers/blob/ad6fd2da0e09233f757bdb0bab0b1ee2b931f33c/src/transformers/trainer.py#L3900-L3905
and
https://github.com/huggingface/transformers/blob/ad6fd2da0e09233f757bdb0bab0b1ee2b931f33c/src/transformers/trainer.py#L5362-L5371

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

in axolotl, remove the hardcoded `average_tokens_across_devices` setting to use the default True, set any of `context_parallel_size` or `tensor_parallel_size` and the training loss will be scaled up by the product of those values.

### Expected behavior

train loss should be similar to the values when using pure DDP.

**Comment by Krish0909** [comment]
I'm looking into this and planning to open a PR with a fix for correctly scaling the loss and token count across only the data parallel group, excluding TP/CP meshes. Will update soon.

**Comment by Rocketknight1** [comment]
cc @sunmarc @S1ro1

**Comment by SunMarc** [comment]
I think we tackled this bit in the cp pr https://github.com/huggingface/transformers/pull/40205. This should solve the issue no @winglian ? cc @S1ro1
